recipes/newrelic/apm/php-agent/php.yml

```diff
@@ -27,16 +29,79 @@ inputVars:
 install:
   version: "3"
 
+  vars:
+    TMP_INSTALL_DIR:
+      sh: mktemp -d /tmp/newrelic-php-agent.XXXXXX
+
   env:
     NEW_RELIC_DISTRIBUTED_TRACING_ENABLED: true
 
   tasks:
     default:
       cmds:
+        - task: create_shared
+        - task: detect_services
         - task: install_packages
         - task: config_newrelic_ini
         - task: finalize_agent_config
         - task: restart_services
+        - task: send_transaction
+        - task: delete_shared
+
+    create_shared:
+      cmds:
+        - rm -rf {{.TMP_INSTALL_DIR}}
+        - mkdir -p {{.TMP_INSTALL_DIR}}
+        - echo "Created temporary directory {{.TMP_INSTALL_DIR}} for installation."
+
+    detect_services:
+      cmds:
+        - echo "Detect services."
+        - |
+          cd {{.TMP_INSTALL_DIR}}
+          rm -f {{.TMP_INSTALL_DIR}}/nonpriv_users.txt
+          rm -f {{.TMP_INSTALL_DIR}}/processes_to_restart.txt
+
+          #
+          # For guided installs, we currently support `php-fpm`, `apache`, and `nginx`.
+          # There is a reason for the order.  According our install script:
+          # If both Apache and FPM are installed, we want FPM to win:
+          # while there are many ways to end up with libapache2-mod-php5 installed,
+          # it's unlikely php5-fpm will be installed unless FPM is actually in use.
+          #
+          declare -a guided_support=("(fpm)" "(httpd|apache2|apache)" "(nginx)")
+
+          current_user=$(whoami)
+
+          for PROCESS in ${guided_support[@]}
+          do
+            nonpriv_user=$(ps -ef | egrep $PROCESS | grep -v grep | grep -v root | head -n1 | awk '{print $1}')
+            priv_user=$(ps -ef | egrep $PROCESS | grep -v grep | head -n1 | awk '{print $1}')
```

Comment:
What guarantees that this entry will be a privileged user? Are there any privileged users other than root?

Reply:
It is the user that starts the process.

Although I know for apache, it should always be root.
Apache is always started as "root" and it then uses "setuid" to spawn children that do the actually processing of requests for the apache user.

The service starts as root in order to do things like bind to reserved ports (e.g. 80 and 443). Then it starts whatever the configured number of processes are, to do the web-server work, and any other tasks, as the defined users. That way requests are being handled by non-privileged processes. You will notice the the parent ID (PPID) is the same for all of the other processes. That idea with be the PID for that one process running as root.

Reply:
Do parent processes always show up before child processes in the output of `ps`? (In my experience, they tend to, but I'm not sure it is a guarantee). If not, the user that started the parent process may not be the user identified by this pipeline.

Reply:
Does `ps` guarantee child process show up after parent processes? (They tend to, in my experience, I just don't know that is guaranteed). If not, you may end up with the less privileged user from a child process entry.

Reply:
Good question.  Using `f` modifier in `ps -ef` I assumed would always list the parent first, but I've changed the flags to a way that guarantees the parent comes first.